### PR TITLE
render.ts - api key auth via process.env

### PR DESCRIPTION
**Pull request by alexfarrill**
from example PR provided

### CODE REVIEW

File: packages/cli/src/pages/api/render.ts [hunk 1]
```diff
@@ -2,19 +2,22 @@ import type { NextApiRequest, NextApiResponse } from "next";
...
 };
 
-export default function handler(
+export default async function handler(
   req: NextApiRequest,
   res: NextApiResponse<Data>
 ) {
   const { templateName, props } = "GET" === req.method ? req.query : req.body;
 
+  if (!(await validateApiKey(req, res))) return;
```
Comment (psugihara) [inline packages/cli/src/pages/api/render.ts:19]: can we put this check behind a config flag? I don't really think we should validate for render by default and don't want to break any existing users.

Maybe 

```
  if (process.env.MAILING_REQUIRE_API_KEY && !(await validateApiKey(req, res))) return;
```

**Review by psugihara — CHANGES_REQUESTED** [review]
looks good, one thing.

Where should we add this to the docs?

File: packages/cli/src/pages/api/render.ts (hunk 1 above)
Comment (alexfarrill) [reply packages/cli/src/pages/api/render.ts:19]: I was thinking the same thing, good suggestion

**Comment by alexfarrill** [comment]
@psugihara have another look, i think it makes sense for render to adopt this behavior if MAILING_API_KEY is set, otherwise not.  and for the validation logic to live separately from the db api key validation (right now they exist as separate features).  wdyt?

**Comment by psugihara** [comment]
Tests look good but I think it's kind of confusing that this api key would work for api/render but not e.g. api/sendMail. With this setup there are 2 kinds of API keys that do 2 different things. I think if I'm setting the api key, it should work as the api key in all contexts. That's simpler conceptually for me.

**Comment by alexfarrill** [comment]
Makes sense. I can flip it to be one function. Do you think the REQUIRE_API_KEY flag should only affect this one type of auth then or apply to all routes and all methods of auth?

**Comment by psugihara** [comment]
Cool. I think we just want REQUIRE_API_KEY to turn on the feature for /api/render (maybe there will be others in the future). Maybe there is a better name for that key.

**Comment by alexfarrill** [comment]
@psugihara updated and added doc

File: packages/web/pages/docs/rest-api.mdx [hunk 2]
```diff
@@ -28,3 +34,10 @@ With the REST API, you can use mailing for email templating even if most of your
 | POST   | /api/lists/[listId]/members            | add a member to a list        |
 | GET    | /api/lists/[listId]/members/[memberId] | get a list member's status    |
 | PATCH  | /api/lists/[listId]/members/[memberId] | update a list member's status |
+
+## API keys
+
+For endpoints that require an API key, the following methods are supported:
+
+- Include the API key in the request headers: `X-API-KEY=...`
+- Include the API key as a query string parameter: `?apiKey=...`
```
Comment (psugihara) [inline packages/web/pages/docs/rest-api.mdx:43]:
```suggestion
## API keys

Mailing uses API keys to protect endpoints that consume limited resources (e.g. sending email with your transport).

To get an API key, either [set up a database](https://www.mailing.run/docs/deploy#adding-a-database) and log in (recommended) or set the `MAILING_API_KEY` environment variable in your deployment environment.

You then pass your API key in each request with one of these methods:

- Include the API key in the request headers: `X-API-KEY=...`
- Include the API key as a query string parameter: `?apiKey=...`


```

File: packages/cli/src/pages/api/render.ts [hunk 3]
```diff
@@ -2,19 +2,27 @@ import type { NextApiRequest, NextApiResponse } from "next";
...
-export default function handler(
+export default async function handler(
   req: NextApiRequest,
   res: NextApiResponse<Data>
 ) {
   const { templateName, props } = "GET" === req.method ? req.query : req.body;
 
+  // if process.env.REQUIRE_API_KEY is set then require a valid api key to be passed in
+  if (process.env.REQUIRE_API_KEY === "true") {
+    if (!(await validateApiKey(req, res))) {
```
Comment (psugihara) [inline packages/cli/src/pages/api/render.ts:21]:
```suggestion
  if (process.env.REQUIRE_API_KEY === "true" && !(await validateApiKey(req, res))) {
```

**Review by psugihara — APPROVED** [review]
looks good. Added to the doc a bit to tell you how to get an API key.

File: packages/web/pages/docs/rest-api.mdx (hunk 2 above)
Comment (alexfarrill) [reply packages/web/pages/docs/rest-api.mdx:43]: Is it clear enough that you need to make up your own API key if you set MAILING_API_KEY?  We might get questions about where to get a key from

Comment (psugihara) [reply packages/web/pages/docs/rest-api.mdx:43]: 👍  edited again for clarity. feel free to edit

Comment (alexfarrill) [reply packages/web/pages/docs/rest-api.mdx:43]: Nice 👍

**Pull request merged**